### ISSUE TITLE
False positive: "<variable>" is not defined Pylance(reportUndefinedVariable)

### ISSUE THREAD

**Issue by shoffmeister** [microsoft/pylance-release#3218]
## Environment data

Pylance language server 2022.8.30 (pyright ed13daf1) starting
Python 3.10.6 (distribution bundled)
Fedora Linux 36

## Code Snippet

Rough sketch:
```python
        if key:  # Enable
            self.key = next((k for k in device.keys if k.key == key), None)
            if self.key:
                self.active = True
                if divertSetting:
                    divertSetting.write_key_value(int(self.key.key), 1)
                device.add_notification_handler(self.name, handler)
                from solaar.ui import status_changed as _status_changed        # <----- this import is good
                self.activate_action()
                _status_changed(device, refresh=True)  # update main window
            else:
                _log.error('cannot enable %s on %s for key %s', self.name, device, key)
        else:  # Disable
            if self.active:
                self.active = False
                if divertSetting:
                    divertSetting.write_key_value(int(self.key.key), 0)
                from solaar.ui import status_changed as _status_changed            # <----- diagnostic on exactly the same import
                _status_changed(device, refresh=True)  # update main window
                try:
                    device.remove_notification_handler(self.name)
                except Exception:
                    if _log.isEnabledFor(_WARNING):
                        _log.warn('cannot disable %s on %s', self.name, device)
                self.deactivate_action()
        return True
```

See actual screenshot below for more context; I am not sure how to approach efficiently trimming down this (to be unfamiliar) code base - my apologies. Pylance implementation knowledge may help to construct a much smaller repro. 

## Repro Steps

* git clone https://github.com/pwr-Solaar/Solaar.git
* open working copy in vscode
* Go to File: lib/logitech_receiver/settings.py

-> get warning: `"status_changed" is not defined Pylance(reportUndefinedVariable)`

You may have to install Python modules to (only) see this diagnostic; in my case, I have the baseline distribution package installed which got me system-wide copies. According to `docs/installation.md` installing the following modules might suffice in a venv:
```
pyudev
psutil
xlib
evdev
yaml
pyyaml
```

## Expected behavior

No warning that `status_changed` is undefined.

## Actual behavior

"status_changed" is not definedPylance[reportUndefinedVariable](https://github.com/microsoft/pylance-release/blob/main/DIAGNOSTIC_SEVERITY_RULES.md#diagnostic-severity-rules)

The screenshot of the snippet below demonstrates the challenge, too: Note the Pylance warning on status_changed at the bottom; the block above mirrors exactly the `from solaar.ui import status_changed as _status_changed` line diagnostic, without any diagnostic. Browsing to that symbol also works from that location. 

![image](https://user-images.githubusercontent.com/3868036/185666389-625d890b-98e5-4da6-a469-3b11af928260.png)

**Comment by erictraut** [comment]
Thanks for posting the quest.

Could you provide a minimal, self-contained code sample that exhibits the issue? It's difficult to repro problems with incomplete code samples and screen shots.

**Comment by shoffmeister** [comment]
I would love to provide better instructions, but I myself am totally unfamiliar with that code base and wouldn't know where to start trimming.

My hope would be that by looking at https://github.com/pwr-Solaar/Solaar/blob/master/lib/logitech_receiver/settings.py#L1403, and with awareness of the implementation details of Pylance, it would be much much easier to trim this down - perhaps even with a breakpoint on a debugger for pylance?

The one thing I can contribute right now is: If the use of status_change on https://github.com/pwr-Solaar/Solaar/blob/master/lib/logitech_receiver/settings.py#L1392 and subsequent lines is removed, then warning disappears.

So pylance seems to get confused by the multiple, identical imports, in the different branches, it would seem. That's the best I can offer for now.

**Comment by shoffmeister** [comment]
For the record, while trying to trim this, I had to restart the Python language server to make this problem appear again, after resetting any modifications.

So in case suddenly the diagnostic disappears - restart the Python language server...

**Comment by rchiodo** [comment]
I can repro opening that same [settings.py](https://github.com/pwr-Solaar/Solaar/blob/master/lib/logitech_receiver/settings.py#L1403):


![Image](https://user-images.githubusercontent.com/19672699/201237108-ac9e3fd8-df5c-473f-88d1-978aee7cfd42.png)


Not sure why it's undefined, but should allow me to debug it.

**Comment by rchiodo** [comment]
This code repros the issue:

```python
def foo():
    if (Bar()):
        from sys import excepthook as _excepthook
        print(_excepthook.__class__)
    else:
        from sys import excepthook as _excepthook
        print(_excepthook.__class__)
```

**Comment by rchiodo** [comment]
![Image](https://user-images.githubusercontent.com/19672699/201443485-7d48c8cd-b9d4-4f50-8bec-07be8ced2ed4.png)

**Comment by rchiodo** [comment]
This issue has been fixed in prerelease version 2022.11.31, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#20221131-16-November-2022-PreRelease)
